src/_P090_MitsubishiHeatPump .ino

```diff
@@ -0,0 +1,698 @@
+  _currentStatus({}),
+  _isConnected(false),
+  _bitrate(2400),
+  _waitForRead(false),
+  _lastSend(0),
+  _lastRecv(millis() - (PACKET_SENT_INTERVAL_MS * 10)),
+  _tempMode(false),
+  _wideVaneAdj(false),
+  _infoMode(0),
+  _isUpdatingSettings(false) {
+
+}
+
+bool P090_data_struct::write(const String& command, const String& value) {
+  if (_isUpdatingSettings) {
+    return false;
+  }
+
+  _isUpdatingSettings = true;
+
+  bool success = false;
+  heatpumpSettings settings = _currentSettings;
+
+  if (command == F("temperature")) {
+    success = string2float(value, settings.temperature);
+  } else if (command == F("power")) {
+    success = lookupValue(POWER_MAP, 2, value.c_str(), settings.power);
+  } else if (command == F("mode")) {
+    success = lookupValue(MODE_MAP, 5, value.c_str(), settings.mode);
+  } else if (command == F("fan")) {
+    success = lookupValue(FAN_MAP, 6, value.c_str(), settings.fan);
+  } else if (command == F("vane")) {
+    success = lookupValue(VANE_MAP, 7, value.c_str(), settings.vane);
+  } else if (command == F("widevane")) {
+    success = lookupValue(WIDEVANE_MAP, 7, value.c_str(), settings.wideVane);
+  }
+
+  if (success && update(settings)) {
+    _currentSettings = settings;
+  } else {
+    success = false;
+  }
+
+  _isUpdatingSettings = false;
+
+  return success;
+}
+
+void P090_data_struct::connect(bool retry) {
+  _isConnected = false;
+  _serial.begin(_bitrate, SERIAL_8E1);
+
+  writePacket(CONNECT, CONNECT_LEN);
+  while(!canRead()) { delay(10); }
+  int packetType = readPacket();
+  if(packetType != RCVD_PKT_CONNECT_SUCCESS && retry){
+	  _bitrate = (_bitrate == 2400 ? 9600 : 2400);
+	  connect(false);
+  }
+}
+
+bool P090_data_struct::read(String& result, bool force) {
+  if (_isUpdatingSettings) {
+    return false;
+  }
+
+  const heatpumpSettings lastSettings = _currentSettings;
+  const heatpumpStatus lastStatus = _currentStatus;
+
+  sync(PACKET_TYPE_DEFAULT);
+
+  if (!_isConnected) {
+    return false;
+  }
+
+  if (force || _currentSettings != lastSettings || _currentStatus.roomTemperature != lastStatus.roomTemperature) {
+    result.reserve(150);
+
+    result = F("{\"roomTemperature\":");
+    result += toString(_currentStatus.roomTemperature, 1);
+    result += F(",\"wideVane\":\"");
+    result += _currentSettings.wideVane;
+    result += F("\",\"power\":\"");
+    result += _currentSettings.power;
+    result += F("\",\"mode\":\"");
+    result += _currentSettings.mode;
+    result += F("\",\"fan\":\"");
+    result += _currentSettings.fan;
+    result += F("\",\"vane\":\"");
+    result += _currentSettings.vane;
+    result += F("\",\"iSee\":");
+    result += boolToString(_currentSettings.iSee);
+    result += F(",\"temperature\":");
+    result += toString(_currentSettings.temperature, 1) + '}';
+
+    return true;
+  }
+
+  return false;
+}
+
+void P090_data_struct::sync(byte packetType) {
+  if((!_isConnected) || (millis() - _lastRecv > (PACKET_SENT_INTERVAL_MS * 10))) {
```

Comment:
Use the functions `timePassedSince` or related functions for checking for timeouts.